StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1028CodeFixProvider.cs

```diff
@@ -0,0 +1,97 @@
+﻿namespace StyleCop.Analyzers.SpacingRules
+{
+    using System;
+    using System.Collections.Immutable;
+    using System.Composition;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CodeActions;
+    using Microsoft.CodeAnalysis.CodeFixes;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.Text;
+
+
+    /// <summary>
+    /// Implements a code fix for <see cref="SA1028NoTrailingWhitespace"/>.
+    /// </summary>
+    /// <remarks>
+    /// <para>To fix a violation of this rule, remove any whitespace at the end of a line of code.</para>
+    /// </remarks>
+    [ExportCodeFixProvider("WhitespaceDiagnosticCodeFixProvider", LanguageNames.CSharp), Shared]
+    public class SA1028CodeFixProvider : CodeFixProvider
+    {
+        private static readonly ImmutableArray<string> FixableDiagnostics =
+            ImmutableArray.Create(SA1028NoTrailingWhitespace.DiagnosticId);
+
+        /// <inheritdoc/>
+        public override ImmutableArray<string> FixableDiagnosticIds => FixableDiagnostics;
+
+        /// <inheritdoc/>
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        /// <inheritdoc/>
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken);
+                var diagnosticSpan = diagnostic.Location.SourceSpan;
+                var trivia = root.FindTrivia(diagnosticSpan.Start, findInsideTrivia: true);
+                if (trivia.Kind() != SyntaxKind.None) // exclude XmlText node till we have support for it.
+                {
+                    context.RegisterCodeFix(
+                        CodeAction.Create(
+                            "Remove trailing whitespace",
+                            ct => RemoveWhitespaceAsync(context.Document, diagnostic, ct)),
+                        diagnostic);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Removes trailing whitespace.
+        /// </summary>
+        /// <param name="document">The document to be changed.</param>
+        /// <param name="diagnostic">The diagnostic to fix.</param>
+        /// <param name="cancellationToken">The cancellation token associated with the fix action.</param>
+        /// <returns>The transformed document.</returns>
+        private static async Task<Document> RemoveWhitespaceAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken);
```

Comment:
I considered it, but decided not to for a few reasons:

1. I expect Roslyn caches the result after the first call, making subsequent calls cheap anyway.
1. When that delegate is invoked, we're applying a code fix, which is a relatively expensive operation anyway, so this operation that was cheap enough to run thousands of times for lots of analyzers can certainly afford to run one more time as part of a user initiated action to apply a code fix.
1. Another parameter on a method makes it that much harder to call, and makes the method that much less helpful to its caller, since the caller has to obtain a value that the helper method could obtain itself.